osquery/TPipe.py

```diff
@@ -0,0 +1,222 @@
+"""This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree. An additional grant
+of patent rights can be found in the PATENTS file in the same directory.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import win32event
+import win32pipe
+import win32file
+import win32api
+import winerror
+import pywintypes
+
+from thrift.transport.TTransport import TTransportBase
+from thrift.transport.TTransport import TTransportException
+from thrift.transport.TTransport import TServerTransportBase
+
+
+class TPipeBase(TTransportBase):
+    """Base class for Windows TPipe transport"""
+
+    def __init__(self):
+        self._handle = None
+
+    def close(self):
+        """
+        Generic close method, as both server and client rely on closing pipes
+        in the same way
+        """
+        if self._handle is not None:
+            win32pipe.DisconnectNamedPipe(self._handle)
+            self._handle = None
+
+
+class TPipe(TPipeBase):
+    """Client for communicating over Named Pipes"""
+
+    def __init__(self, pipeName, timeout=5, maxAttempts=5):
+        """
+        Initialize a TPipe client
+
+        @param pipeName(string)  The path of the pipe to connect to.
+        @param timeout(int)  The time to wait for a named pipe connection.
+        @param maxAttempts(int)  Maximum number of connection attempts.
+        """
+        self._handle = None
+        self._pipeName = pipeName
+        self._timeout = timeout
+        self._maxConnAttempts = maxAttempts
+
+    def setHandle(self, h):
+        self._handle = h
+
+    def isOpen(self):
+        return self._handle is not None
+
+    def open(self):
+        if self.isOpen():
+            raise TTransportException(TTransportException.ALREADY_OPEN)
+
+        h = None
+        conns = 0
+        while conns < self._maxConnAttempts:
+            try:
+                h = win32file.CreateFile(
+                    self._pipeName,
+                    win32file.GENERIC_READ | win32file.GENERIC_WRITE, 0, None,
+                    win32file.OPEN_EXISTING, win32file.FILE_FLAG_OVERLAPPED,
+                    None)
+            except Exception as e:
+                if e[0] != winerror.ERROR_PIPE_BUSY:
+                    raise TTransportException(
+                        TTransportException.NOT_OPEN,
+                        'Failed to open connection to pipe: {}'.format(e))
+
+            # Successfully connected, break out.
+            if h is not None and h.handle != winerror.ERROR_INVALID_HANDLE:
```

Comment:
it's fine, nit it's better to check the type of the structure you look for in `h`. 

This also could be `if not h and h.handle ...`